File: app/src/main/java/com/example/uscatterbrain/network/InputStreamObserver.java
package com.example.uscatterbrain.network;

import android.util.Log;

import java.io.IOException;

import io.reactivex.Observer;
import io.reactivex.disposables.Disposable;

public class InputStreamObserver extends InputStreamCallback implements Observer<byte[]> {

    @Override
    public void onSubscribe(Disposable d) {
        disposable = d;
    }

    @Override
    public void onNext(byte[] bytes) {
        if (!closed) {
            acceptBytes(bytes);
        }
    }

    @Override
    public void onError(Throwable e) {
        throwable = e;
        try {
            close();
        } catch (IOException ignored) {

        }
    }

    @Override
    public void onComplete() {
        complete();
    }
}

File: ser/src/main/java/net/ballmerlabs/scatterproto/InputStreamObserver.kt
package net.ballmerlabs.scatterproto

import io.reactivex.Observer
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.disposables.Disposable
import java.io.IOException

/**
 * serves as a bridge between an observable and an inputstream
 * using a static sized buffer to hold unread data
 */
class InputStreamObserver(capacity: Int) : InputStreamCallback(capacity), Observer<ByteArray>, Disposable {

    val disp = CompositeDisposable()

    override fun onSubscribe(d: Disposable) {
        disp.add(d)
    }

    override fun dispose() {
        disp.dispose()
    }

    override fun isDisposed(): Boolean {
        return disp.isDisposed
    }

    override fun onNext(bytes: ByteArray) {
        if (!closed) {
            acceptBytes(bytes)
        }
    }

    override fun onError(e: Throwable) {
        throwable = e
        try {
            close()
        } catch (ignored: IOException) {
        }
    }

    override fun onComplete() {
    }
}